oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2TokenIntrospectionResponse.java

```diff
@@ -0,0 +1,270 @@
+			return this;
+		}
+
+		/**
+		 * Populates the 'token_type' field.
+		 *
+		 * @param tokenType {@link TokenType} indicating the type of the token
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder tokenType(OAuth2AccessToken.TokenType tokenType) {
+			this.addParam(TOKEN_TYPE, tokenType.getValue());
+			return this;
+		}
+
+		/**
+		 * Populates the 'exp' (Expiration Time) field.
+		 *
+		 * @param expirationTime {@link Instant} indicating when this token will expire
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder expirationTime(Instant expirationTime) {
+			this.addParam(EXP, expirationTime.getEpochSecond());
+			return this;
+		}
+
+		/**
+		 * Populates the 'iat' (Issued At) field.
+		 *
+		 * @param issuedAt {@link Instant} indicating when this token was originally issued
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder issuedAt(Instant issuedAt) {
+			this.addParam(IAT, issuedAt.getEpochSecond());
+			return this;
+		}
+
+		/**
+		 * Populates the 'nbf' (Not Before) field.
+		 *
+		 * @param notBefore {@link Instant} indicating when this token is not to be used before
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder notBefore(Instant notBefore) {
+			this.addParam(NBF, notBefore.getEpochSecond());
+			return this;
+		}
+
+		/**
+		 * Populates the 'sub' (Subject) field.
+		 *
+		 * @param subject usually a machine-readable identifier of the resource owner who authorized this token
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder subject(String subject) {
+			this.addParam(SUB, subject);
+			return this;
+		}
+
+		/**
+		 * Populates the 'aud' (Audience) field.
+		 *
+		 * @param audience service-specific string identifier or list of string identifiers representing the intended audience for this
+		 * token
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder audience(List<String> audience) {
+			this.addParam(AUD, audience);
+			return this;
+		}
+
+		/**
+		 * Populates the 'iss' (Issuer) field.
+		 *
+		 * @param issuer of this token
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder issuer(String issuer) {
+			this.addParam(ISS, issuer);
+			return this;
+		}
+
+		/**
+		 * Populates the 'jti' (JWT ID) field.
+		 *
+		 * @param jwtId identifier for the token
+		 * @return the {@link Builder} for further configurations
+		 */
+		public Builder jwtId(String jwtId) {
+			this.addParam(JTI, jwtId);
+			return this;
+		}
+
+		/**
+		 * Build the {@link OAuth2TokenIntrospectionResponse}
+		 *
+		 * @return The constructed {@link OAuth2TokenIntrospectionResponse}
+		 */
+		public OAuth2TokenIntrospectionResponse build() {
+			Map<String, Object> responseParams = this.params.entrySet().stream()
+					.filter(entry -> SUPPORTED_FIELDS.contains(entry.getKey()))
```

Comment:
The spec indicates:

> Specific implementations MAY extend this structure with their own
   service-specific response names as top-level members of this JSON
   object.

Thus we might want to remove this filter if we want to retrieve custom claims.
For the time being, I limited the response to avoid leaking entries we might have in the token metadata claims.